src/diffusers/loaders/single_file.py

```diff
@@ -252,71 +332,183 @@ def from_single_file(cls, pretrained_model_link_or_path, **kwargs):
-        )
+            # We might want to deprecate this behavior as it is brittle and can lead to errors
+            component_types = pipeline_class._get_signature_types()
+            expected_modules, optional_kwargs = pipeline_class._get_signature_keys(cls)
+            component_types = {
+                k: v for k, v in component_types.items() if k not in pipeline_class._optional_components
+            }
+
+            config_dict = _map_component_types_to_config_dict(component_types)
+            config_dict["_class_name"] = pipeline_class.__name__
+
+        else:
+            if not os.path.isdir(default_pretrained_model_name_or_path):
+                # Provided config is a repo_id
+                if default_pretrained_model_name_or_path.count("/") > 1:
+                    raise ValueError(
+                        f'The provided pretrained_model_name_or_path "{default_pretrained_model_name_or_path}"'
+                        " is neither a valid local path nor a valid repo id. Please check the parameter."
+                    )
+                config_file = hf_hub_download(
+                    default_pretrained_model_name_or_path,
+                    filename=cls.config_name,
+                    cache_dir=cache_dir,
+                    revision=revision,
+                    proxies=proxies,
+                    force_download=force_download,
+                    resume_download=resume_download,
+                    local_files_only=local_files_only,
+                    token=token,
+                )
 
-        expected_modules, optional_kwargs = cls._get_signature_keys(pipeline_class)
+                config_dict = pipeline_class._dict_from_json_file(config_file)
+
+            else:
+                # Provide config is a path to a local directory attempt to load directly.
+                config_dict = pipeline_class.load_config(default_pretrained_model_name_or_path)
+
+        #   pop out "_ignore_files" as it is only needed for download
+        config_dict.pop("_ignore_files", None)
+
+        expected_modules, optional_kwargs = pipeline_class._get_signature_keys(cls)
         passed_class_obj = {k: kwargs.pop(k) for k in expected_modules if k in kwargs}
         passed_pipe_kwargs = {k: kwargs.pop(k) for k in optional_kwargs if k in kwargs}
 
-        model_type = kwargs.pop("model_type", None)
-        image_size = kwargs.pop("image_size", None)
-        load_safety_checker = (kwargs.pop("load_safety_checker", False)) or (
-            passed_class_obj.get("safety_checker", None) is not None
-        )
+        init_dict, unused_kwargs, _ = pipeline_class.extract_init_dict(config_dict, **kwargs)
+
+        from diffusers import pipelines
+
+        init_kwargs = {
+            k: init_dict.pop(k) for k in optional_kwargs if k in init_dict and k not in cls._optional_components
+        }
+        init_kwargs = {**init_kwargs, **passed_pipe_kwargs}
+
+        # remove `null` components
```

Comment:
Could you give an example situation that might produce null components?

Reply:
We use this in `from_pretrained` as well 
https://github.com/huggingface/diffusers/blob/26a7851e1e0b18da746d6ae80bb105050f7187e0/src/diffusers/pipelines/pipeline_utils.py#L802

It filters out components that are not loadable modules e.g. in SDXL we have arguments like `force_zeros_for_empty_prompt` or if we pass `None` for a components e.g `safety_checker=None`